ir/src/constraints/graph.rs

```diff
@@ -58,130 +53,271 @@ impl AlgebraicGraph {
+    /// return the other expression.
+    pub(super) fn merge_equal_exprs(
+        &mut self,
+        lhs: &ExprDetails,
+        rhs: &ExprDetails,
+    ) -> Result<ExprDetails, SemanticError> {
+        let node_index = self.insert_op(Operation::Sub(lhs.root_idx(), rhs.root_idx()));
+        let trace_segment = lhs.trace_segment().max(rhs.trace_segment());
+        let domain = lhs.domain().merge(&rhs.domain())?;
+
+        Ok(ExprDetails::new(node_index, trace_segment, domain))
+    }
 
     /// Adds the expression to the graph and returns the [ExprDetails] of the constraint.
     /// Expressions are added recursively to reuse existing matching nodes.
     pub(super) fn insert_expr(
         &mut self,
         symbol_table: &SymbolTable,
-        expr: Expression,
+        expr: &Expression,
         variable_roots: &mut VariableRoots,
+        default_domain: ConstraintDomain,
     ) -> Result<ExprDetails, SemanticError> {
         match expr {
             Expression::Const(value) => {
-                let node_index = self.insert_op(Operation::Constant(ConstantValue::Inline(value)));
-                Ok((node_index, DEFAULT_SEGMENT, DEFAULT_DOMAIN))
+                let node_index = self.insert_op(Operation::Constant(ConstantValue::Inline(*value)));
+                Ok(ExprDetails::new(
+                    node_index,
+                    DEFAULT_SEGMENT,
+                    default_domain,
+                ))
             }
             Expression::Elem(Identifier(ident)) => {
-                self.insert_symbol_access(symbol_table, &ident, variable_roots)
-            }
-            Expression::VectorAccess(vector_access) => {
-                self.insert_vector_access(symbol_table, &vector_access, variable_roots)
-            }
-            Expression::MatrixAccess(matrix_access) => {
-                self.insert_matrix_access(symbol_table, &matrix_access, variable_roots)
+                self.insert_symbol_access(symbol_table, ident, variable_roots, default_domain)
             }
+            Expression::VectorAccess(vector_access) => self.insert_vector_access(
+                symbol_table,
+                vector_access,
+                variable_roots,
+                default_domain,
+            ),
+            Expression::MatrixAccess(matrix_access) => self.insert_matrix_access(
+                symbol_table,
+                matrix_access,
+                variable_roots,
+                default_domain,
+            ),
             Expression::Rand(index) => {
                 // constraint target for random values defaults to the second trace segment.
                 // TODO: make this more general, so random values from further trace segments can be
                 // used. This requires having a way to describe different sets of randomness in
                 // the AirScript syntax.
                 let trace_segment = 1;
```

Comment:
nit: Should we define an alias `AUX_TRACE_SEGMENT`?

Reply:
Yeah, I think that makes sense. In the longer run we should make this more general, but that's very low priority, so we probably won't get there for a while